text/0626-invoke-helper.md

```diff
@@ -0,0 +1,475 @@
+being that a helper manager has been associated with it via the
+`setHelperManager` API.
+
+#### `argsGetter`
+
+This is an optional function that produces the arguments to the helper. The
+function receives the parent context as an argument, and must return an object
+with a `positional` property that is an array and/or a `named` property that is
+an object.
+
+This getter function will be _autotracked_ when it is run, so the process of
+retrieving the arguments is autotracked. If any of the values used to create the
+arguments object change, the helper will be updated, just like in templates.
+
+### Return Value
+
+The function returns an instance of the helper. The public API of the instance
+consists of a `value` property, which will internally be implemented as a getter
+that triggers the proper lifecycle hooks on the helper and returns its value, if
+it has a value. If it does not, then the helper will do nothing when `value` is
+accessed and it will always return `undefined`.
+
+If the helper has a scheduled effect, there is no public API for users to access
+the effect or run it eagerly. It will run as scheduled, until the helper is
+destroyed.
+
+Using `destroy()` from the destroyables API on the helper instance will trigger
+its destruction early. Users can do this to clean up a helper before the parent
+context is destroyed.
+
+### Effect Helper Timing Semantics
+
+Standard helpers that return a value will only be updated when they are used,
+either in JavaScript or in the template. The args getter and the relevant helper
+manager lifecycle hooks will be called when the `value` property on the helper
+is used.
+
+Side-effecting helpers, by contrast, run their updates specifically when
+scheduled. When introduced by the Helper Manager RFC, there was no relative
+ordering specified in the scheduling of side-effecting helpers, because there
+was no way for them to have _children_, and we don't generally specify ordering
+of siblings. With this RFC, it will be possible to invoke a side-effecting
+helper within another side-effecting helper, so they will be able to have
+children for the first time.
+
+This RFC proposes modifying the Helper Manager RFC to specify that the
+`runEffect` hook of a helper always runs _after_ the `runEffect` hooks of its
+children. This mirrors the timing semantics of modifier hooks in templates.
+
+### Ergonomics
+
+This is a low-level API for invoking helpers and creating instances. The API is
+meant to be functional, but not particularly readable or ergonomic. This API can
+be wrapped with higher level, more ergonomic APIs in the ecosystem, until we're
+sure what the final API should be.
+
+## How we teach this
+
+This API is meant to be a low-level primitive which will eventually be replaced
+with higher level wrappers, possibly decorators, that will be much easier to use
+and recommend to average app developers. As such, it will only be taught through
+API documentation.
```

Comment:
I fear it's a little bit more complex than this. Helpers are currently reached as *template* helpers. This proposal decouples them from the template. They could be invoked everywhere. It might not be important to teach the low-level API to invoke them. But it would be important to reflect in the guides that they could also be used outside of template context.

Reply:
Yes, I do agree there. I think we should update the guides to reflect this once we have a solid community example, similar to what we did with `ember-modifier`. I'll add that to the RFC.